spacy/language.py

```diff
@@ -1295,27 +1296,27 @@ def evaluate(
             kwargs = dict(scorer_cfg)
             kwargs.setdefault("nlp", self)
             scorer = Scorer(**kwargs)
-        texts = [eg.reference.text for eg in examples]
-        docs = [eg.predicted for eg in examples]
+        # reset annotation in predicted docs and time tokenization
         start_time = timer()
-        # tokenize the texts only for timing purposes
         if not hasattr(self.tokenizer, "pipe"):
-            _ = [self.tokenizer(text) for text in texts]  # noqa: F841
+            for eg in examples:
+                eg.predicted = self.tokenizer(eg.reference.text)
         else:
-            _ = list(self.tokenizer.pipe(texts))  # noqa: F841
+            for doc, eg in zip(
+                self.tokenizer.pipe(eg.reference.text for eg in examples), examples
+            ):
+                eg.predicted = doc
+        # apply all pipeline components
         for name, pipe in self.pipeline:
             kwargs = component_cfg.get(name, {})
             kwargs.setdefault("batch_size", batch_size)
-            docs = _pipe(docs, pipe, kwargs)
-        # iterate over the final generator
-        if len(self.pipeline):
-            docs = list(docs)
+            for doc, eg in zip(
```

Comment:
I don't think this construction will work if `examples` is a generator. You can change the `_pipe` function, or you can use `itertools.tee`.

The issue is you're going to consume `examples` in one of the iterations, and not have it available for both. You could also add a batching loop as another solution.

Reply:
It's converted to a list at the top already, because otherwise nothing would work here in terms of storing the results as `eg.predicted`.

Reply:
(I don't think the list of example objects is going to be that large in memory?)